feat(json): support pretty-formatting with `vim.json.encode()`

arkriny opened the pull request:
## Problem:
There is no straightforward way to pretty-print objects as JSON.
The existing `vim.inspect` outputs LON.

## Solution:
Introduce an `indent` option for `vim.json.encode()` which enables human-readable output with configurable indentation.

Adapts PR to upstream: https://github.com/openresty/lua-cjson/pull/114.

CC @echasnovski for the original proposal and draft implementation.

justinmk commented:
Wasn't the main goal to sort by key name? Don't see that mentioned in the diff.

echasnovski commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,15 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
+                • pretty: (table) Table with keys:
+                  • enable: (boolean) (default false) Whether to enable
+                    pretty-formatting.
+                  • indent: (string) (default " ") String used for each
```
This ate a single space during document generation (`" "`, but should be `"  "`). Maybe needs escaping with `` ` ``.

skewb1k commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,15 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
                 • escape_slash: (boolean) (default false) Escape slash
                   characters "/" in string values.
+                • pretty: (table) Table with keys:
+                  • enable: (boolean) (default false) Whether to enable
```
I'm not fully sure about this API either. Having an "enable" flag makes it easier to use without having to provide every option explicitly. But it's also possible to avoid "enable" by using non-empty defaults, so that `{ pretty = {} }` would work. I haven't seen this pattern used in existing Lua APIs though, maybe there are some concerns with it. I think the best we can do is stay consistent either way.

mfussenegger replied:
>  but it does not let you change newline characters.

Is there a legitimate use-case for wanting indentation but not newlines?

skewb1k replied:
> Is there a legitimate use-case for wanting indentation but not newlines?

Probably not, but there could be cases where someone wants a newline character other than `\n`. We can't know for sure.

I think we shouldn’t strip customizability just to make the configuration slightly less verbose for what we assume is the more common use case.

MariaSolOs commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,15 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
                 • escape_slash: (boolean) (default false) Escape slash
                   characters "/" in string values.
+                • pretty: (table) Table with keys:
+                  • enable: (boolean) Whether to enable pretty-formatting.
```
Can't we remove `enable` here and instead make `pretty` optional, such that if the argument is provided pretty formatting is implied?

skewb1k replied:
We can, that seems like a good idea.

See [comment](https://github.com/neovim/neovim/pull/35424#discussion_r2293165588)

MariaSolOs commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,15 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
+                  • enable: (boolean) Whether to enable pretty-formatting.
+                    Defaults to false.
+                  • indent: (string) String used for each indentation level.
+                    Defaults to `"  "`.
```
```suggestion
                    Defaults to `"  "` (2 spaces).
```

MariaSolOs commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,15 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
       • {opts}  (`table<string,any>?`) Options table with keys:
                 • escape_slash: (boolean) (default false) Escape slash
                   characters "/" in string values.
+                • pretty: (table) Table with keys:
```
What kind of error do we show if the provided formatting options result in invalid JSON?

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,24 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
+                • pretty: (table|boolean) (default false) Enables
+                  pretty-formatting when provided or `true`. Supports the
+                  following keys:
+                  • indent: (integer|string) (default 2) Controls how nested
```
can drop the integar variant. it's trivial to convert a number to spaces, `(' '):rep(4)`

justinmk replied:
It's mostly about simplifying the interface and docs. When the docs get nested like this, we need to ask why.

The door is always open in the future to add more variations like this. So we don't need to have all these features right now.

This PR enables something that was previously impossible. There is no need to also add bells and whistles in this iteration.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,24 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
+                    • If a string: that string is used for the indentation.
+                      Must contain only spaces, tabs, line feeds (LF), or
+                      carriage returns (CR).
+                  • item_separator: (string) (default "\n") Separator inserted
```
```suggestion
                  • itemsep: (string) (default "\n") Separator inserted
```

justinmk commented:
@echasnovski do you actually need `key_separator` / `item_separator` ? those seem rather special-case, and if that's really needed, perhaps instead it should just be a `format` callback so the user can control the formatting, similar to `replacer` in [JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify).

mike325 commented on runtime/doc/lua.txt:
```diff
@@ -3375,6 +3375,24 @@ vim.json.encode({obj}, {opts})                             *vim.json.encode()*
+                • pretty: (table|boolean) (default false) Enables
+                  pretty-formatting when provided or `true`. Supports the
+                  following keys:
+                  • indent: (integer|string) (default 2) Controls how nested
```
I don't know how much complicated would it be but I think taking the value from something like `vim.filetype.get_option()` might be a slightly more sensible approach, this way it will be formatted following the user's options.

skewb1k replied:
I don't think it should be baked into `vim.json`; it should be handled by the caller. This way, the output with default values remains consistent and does not depend on state. It also avoids assuming that the output is written to a file, which only the caller can know.

skewb1k commented:
Should we allow `indent = true` to default to 2 spaces?

justinmk commented:
> Should we allow `indent = true` to default to 2 spaces?

as hinted by https://github.com/neovim/neovim/pull/35424#discussion_r2299596838 , we should be avoiding multi-type parameters, not adding more.

arkriny committed:
feat(json): support pretty-formatting with `vim.json.encode()`

Problem:
There is no straightforward way to pretty-print objects as JSON.
The existing `vim.inspect` outputs LON.

Solution:
Introduce an `indent` option for `vim.json.encode()` which enables
human-readable output with configurable indentation.

Adapts PR to upstream: openresty/lua-cjson#114.

justinmk commented on src/cjson/lua_cjson.c:
```diff
@@ -436,6 +453,41 @@ static int json_cfg_encode_keep_buffer(lua_State *l)
 }
 */
 
+/* Returns nonzero if the string consists only of allowed whitespace characters */
```
why does this api care what the user decides they want to "indent" with?

skewb1k replied:
For the same reason it cares about the first argument being deserializable: it ensures the successful result is always valid JSON. We should treat indent as user input as well. The difference is that the first argument is obviously could be user input, and nobody wants to validate it manually, while `indent` is less likely to come from user input and easier to validate.
Still, the performance cost of checking it is low, and validation can be removed later without breaking compatibility. In contrast, if we didn't validate and tried to add it later, that would be a backward-incompatible change.

justinmk reviewed (approved):
LGTM, let's wait a couple days for a reply to https://github.com/openresty/lua-cjson/pull/114 then can decide how to proceed
